Framework/Algorithms/src/PolarizationCorrections/PolarizerEfficiency.cpp

```diff
@@ -0,0 +1,201 @@
+                  "File name or path for the output to be saved to.");
+}
+
+/**
+ * Tests that the inputs are all valid
+ * @return A map containing the incorrect workspace
+ * properties and an error message
+ */
+std::map<std::string, std::string> PolarizerEfficiency::validateInputs() {
+  std::map<std::string, std::string> errorList;
+  // Check input workspaces.
+  const WorkspaceGroup_sptr inputWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
+  if (inputWorkspace == nullptr) {
+    errorList[PropertyNames::INPUT_WORKSPACE] = "The input workspace is not a workspace group.";
+    return errorList;
+  }
+
+  const MatrixWorkspace_sptr analyserWs = getProperty(PropertyNames::ANALYSER_EFFICIENCY);
+  if (analyserWs == nullptr) {
+    errorList[PropertyNames::ANALYSER_EFFICIENCY] = "The analyser efficiency workspace is not a MatrixWorkspace.";
+    return errorList;
+  }
+  if (analyserWs->getNumberHistograms() != 1) {
+    errorList[PropertyNames::ANALYSER_EFFICIENCY] = "All workspaces must contain a single histogram.";
+  }
+
+  auto const &inputWsCount = inputWorkspace->size();
+  if (inputWsCount < 2 || inputWsCount > 4) {
+    errorList[PropertyNames::INPUT_WORKSPACE] =
+        "The input group workspace must have at least two periods corresponding to the spin configurations.";
+  } else {
+    for (size_t i = 0; i < inputWsCount; ++i) {
+      const MatrixWorkspace_sptr stateWs = std::dynamic_pointer_cast<MatrixWorkspace>(inputWorkspace->getItem(i));
+      Unit_const_sptr unit = stateWs->getAxis(0)->unit();
+      if (unit->unitID() != "Wavelength") {
+        errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be in units of Wavelength.";
+      }
+      if (stateWs->getNumberHistograms() != 1) {
+        errorList[PropertyNames::INPUT_WORKSPACE] = "All workspaces must contain a single histogram.";
+      }
+    }
+  }
+
+  // Check outputs.
+  auto const &outputWs = getPropertyValue(PropertyNames::OUTPUT_WORKSPACE);
+  auto const &outputFile = getPropertyValue(PropertyNames::OUTPUT_FILE_PATH);
+  if (outputWs.empty() && outputFile.empty()) {
+    errorList[PropertyNames::OUTPUT_FILE_PATH] = "Either an output workspace or output file must be provided.";
+    errorList[PropertyNames::OUTPUT_WORKSPACE] = "Either an output workspace or output file must be provided.";
+  }
+
+  return errorList;
+}
+
+void PolarizerEfficiency::exec() { calculatePolarizerEfficiency(); }
+
+void PolarizerEfficiency::calculatePolarizerEfficiency() {
+  // First we extract the individual workspaces corresponding to each spin configuration from the group workspace
+  const WorkspaceGroup_sptr &groupWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
+  const auto spinConfigurationInput = getPropertyValue(PropertyNames::SPIN_STATES);
+
+  const auto &t01Ws = PolarizationCorrectionsHelpers::workspaceForSpinState(groupWorkspace, spinConfigurationInput,
```

Comment:
I think we need some validation to check that `01` and `00` are included in the spin state configuration input, even if there are more than two configurations supplied. I tested using an input workspace group of 4 workspaces and with spin configuration input `11, 00` and it ran through without error. The calculated efficiencies were different compared to when I set the spin configuration input to `01, 00`. I think we possibly want to validate that whatever number of spin configurations has been passed in matches the number of workspaces in the group. But at the very least, I think we need to check that we have the required `01` and `00` for the calculation.

In my first invalid test I was surprised not to get an error from `PolarizationCorrectionsHelpers::workspaceForSpinState`, because it's looking for spin state `01` that wasn't passed in, so that could do with a look.

Reply:
Turns out the source of this is due to the way the Helper function works. If it fails to find a match for a given spin state (01, say) it returns the index 1 beyond the last index in the list of spin states. This means that if the sizes do not match, the helper can return an invalid index for the spin states that _is_ valid for the workspace group (provided it is larger). I'll make a fix for both issues and validate for both, as it's just confusing otherwise.